File: code/cpp/stack_heap.cpp
#define unused __attribute ((unused))
int main(unused int argc, unused char **argv)
// BEGIN_CODE
{
    // single char allocated on the stack
    char a = 'a';

    // single char allocated on the heap
    // char pointer allocated on the stack
    char *b = new char('b');

    // array of 23 chars allocated on the heap
    // char pointer allocated on the stack
    char *c = new char[23];


    // heap allocated char is freed
    delete b;

    // heap allocated array is freed
    delete[] c;
}
// END_CODE

File: code/rust/stack_heap.rs
use std::mem;

#[allow(unused_variables)]
fn main()
// BEGIN_CODE
{
    // single char allocated on the stack
    let a: char = 'a';

    // single char allocated on the heap
    // 
    let b = Box::new('b');

    // array of 23 chars allocated on the heap
    // char pointer allocated on the stack
    let c = Box::new(['c'; 23]);

    // heap allocated memory is freed when the
    // variable goes out of scope
    // however, you can force freeing by using
    // the mem::drop method
    mem::drop(b);
    mem::drop(c);
}
// END_CODE

